### PR TITLE
Unify task (un)blocking in physical operators

### PR DESCRIPTION
**Pull request by lnkuiper**
... and add functionality to _prevent_ tasks from being blocked, e.g., when we have a streaming `LIMIT` on top of a blocking source operator. This has caused us some deadlock issues in the past, which this PR fixes, and hopefully prevents from happening in the future.

### CODE REVIEW

File: src/execution/operator/aggregate/physical_window.cpp [hunk 1]
```diff
@@ -76,17 +76,17 @@ class WindowHashGroup {
...
+			break;
 		case WindowGroupStage::FINALIZE:
 			if (finalized == blocks) {
 				stage = WindowGroupStage::GETDATA;
 				return true;
 			}
-			return false;
+			break;
 		default:
-			// never block in GETDATA
```
Comment (Mytherin) [inline src/execution/operator/aggregate/physical_window.cpp:87]: We're still never blocking here - and if I modify this to block then I get deadlocks again. Perhaps we should leave this as-is and leave the comment as well?

File: src/execution/operator/persistent/physical_batch_insert.cpp [hunk 2]
```diff
@@ -475,11 +481,11 @@ SinkResultType PhysicalBatchInsert::Sink(ExecutionContext &context, DataChunk &c
 			// execute tasks while we wait (if any are available)
 			ExecuteTasks(context.client, gstate, lstate);
 
-			lock_guard<mutex> l(memory_manager.GetBlockedTaskLock());
+			auto guard = memory_manager.Lock();
 			if (!memory_manager.IsMinimumBatchIndex(batch_index)) {
 				//  we are not the minimum batch index and we have no memory available to buffer - block the task for
 				//  now
-				memory_manager.BlockTask(input.interrupt_state);
+				memory_manager.BlockTask(guard, input.interrupt_state);
```
Comment (Mytherin) [inline src/execution/operator/persistent/physical_batch_insert.cpp:488]: The return value here seems to be unused

File: src/execution/operator/join/physical_hash_join.cpp [hunk 3]
```diff
@@ -1192,15 +1190,15 @@ SourceResultType PhysicalHashJoin::GetData(ExecutionContext &context, DataChunk
...
 			if (gstate.TryPrepareNextStage(sink) || gstate.global_stage == HashJoinSourceStage::DONE) {
-				for (auto &state : gstate.blocked_tasks) {
-					state.Callback();
-				}
-				gstate.blocked_tasks.clear();
+				gstate.UnblockTasks(guard);
 			} else {
-				gstate.blocked_tasks.push_back(input.interrupt_state);
-				return SourceResultType::BLOCKED;
+				if (gstate.BlockTask(guard, input.interrupt_state)) {
```
Comment (Mytherin) [inline src/execution/operator/join/physical_hash_join.cpp:1197]: Perhaps we can add a `BlockSource` and `BlockSink` that directly return this, so we can do `return gstate.BlockSource(...)` here?

File: src/execution/operator/persistent/physical_batch_copy_to_file.cpp [hunk 4]
```diff
@@ -163,11 +163,11 @@ SinkResultType PhysicalBatchCopyToFile::Sink(ExecutionContext &context, DataChun
 		FlushBatchData(context.client, gstate);
 
 		if (!memory_manager.IsMinimumBatchIndex(batch_index) && memory_manager.OutOfMemory(batch_index)) {
-			lock_guard<mutex> l(memory_manager.GetBlockedTaskLock());
+			auto guard = memory_manager.Lock();
 			if (!memory_manager.IsMinimumBatchIndex(batch_index)) {
 				// no tasks to process, we are not the minimum batch index and we have no memory available to buffer
 				// block the task for now
-				memory_manager.BlockTask(input.interrupt_state);
+				memory_manager.BlockTask(guard, input.interrupt_state);
```
Comment (Mytherin) [inline src/execution/operator/persistent/physical_batch_copy_to_file.cpp:170]: The return value here seems to be unused

File: src/execution/radix_partitioned_hashtable.cpp [hunk 5]
```diff
@@ -678,7 +673,7 @@ SourceResultType RadixHTGlobalSourceState::AssignTask(RadixHTGlobalSinkState &si
 	case AggregatePartitionState::FINALIZE_IN_PROGRESS:
 		lstate.task = RadixHTSourceTaskType::SCAN;
 		lstate.scan_status = RadixHTScanStatus::INIT;
-		partition.blocked_tasks.push_back(interrupt_state);
+		partition.BlockTask(partition_guard, interrupt_state);
```
Comment (Mytherin) [inline src/execution/radix_partitioned_hashtable.cpp:676]: The return value here seems to be unused

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks great. Some minor comments:

File: src/execution/operator/aggregate/physical_window.cpp (hunk 1 above)
Comment (Mytherin) [reply src/execution/operator/aggregate/physical_window.cpp:87]: For future reference - the way I tested the deadlocks is with this diff:

```diff
diff --git a/benchmark/include/benchmark.hpp b/benchmark/include/benchmark.hpp
index ece2dda3d3..acd2f3d887 100644
--- a/benchmark/include/benchmark.hpp
+++ b/benchmark/include/benchmark.hpp
@@ -28,7 +28,7 @@ struct BenchmarkState {
 //! The base Benchmark class is a base class that is used to create and register
 //! new benchmarks
 class Benchmark {
-       constexpr static size_t DEFAULT_NRUNS = 5;
+       constexpr static size_t DEFAULT_NRUNS = 500;
        Benchmark(Benchmark &) = delete;
 
 public:
diff --git a/src/execution/operator/aggregate/physical_window.cpp b/src/execution/operator/aggregate/physical_window.cpp
index 9435c24de7..9f9341fcdb 100644
--- a/src/execution/operator/aggregate/physical_window.cpp
+++ b/src/execution/operator/aggregate/physical_window.cpp
@@ -84,7 +84,7 @@ public:
                        }
                        break;
                default:
-                       return true;
+                       break;
                }
                return false;
        }

```

Then running:

```
build/reldebug/benchmark/benchmark_runner benchmark/tpcds/sf1/q49.benchmark
```

Which gets stuck for me spuriously in one of the hundreds of runs.

**Comment by lnkuiper** [comment]
Thanks for the review! Looks much cleaner now

**Pull request merged**